### PR TITLE
Assembler simplification (pending state removal)

### PR DESCRIPTION
**Pull request by gzanitti**
This PR simplifies the operation of the assembler, mainly by removing the need to maintain a pending state and to return the assembler to its original state in certain situations (which caused issue #124).

The main changes include
- The creation of two structs, PendingMacro and PendingLabel, which store information about the macros and labels used before they were defined, as well as their relative position in the code.
- Some push operations have been changed to take an option<usize> as parameter, and if this is not None, to act more like an insert than a push (useful for example to insert the calls to the pending macros).

Also, issue #124 has been fixed and the corresponding test added.

All tests should pass successfully. However, I would like to test the new assembler a bit more by writing some code (e.g. following @lightclient 's  recommendation and writing an ERC-20 implementation using ETK) before merging it.

In case you don't find it necessary, it can be merged right now.

### CODE REVIEW

**Comment by gzanitti** [comment]
@lightclient @SamWilsn the assembler rewrite is ready. If anyone has time during the week to review the code, I'll be happy to address your comments :)

P.S: I'm tagging you two because you're the two people I've been talking to about the assembler rewrite. If you don't have time, feel free to ignore this message.

**Comment by SamWilsn** [comment]
@gzanitti we're the correct people to tag for sure! I'm swamped for the rest of the week, but I start my vacation next week so I should finally have some time for ETK :tada:

**Comment by gzanitti** [comment]
> @gzanitti we're the correct people to tag for sure! I'm swamped for the rest of the week, but I start my vacation next week so I should finally have some time for ETK 🎉

That's great! :). Anyway, today I was playing around with an ERC20 implementation I was working on (basically, moving different parts of the code to macros, adding includes/imports) and I found a bug. Which I thought I had fixed a few hours ago, but it turns out that the new change breaks some "push label"s haha. 

I'll try to solve it in the next few days, but I'm starting to think it might need a bit of a major rewrite. I'll keep you posted.

**Comment by SamWilsn** [comment]
That might be the best approach. Keep the pest parser and scrap the rest.

**Comment by gzanitti** [comment]
> That might be the best approach. Keep the pest parser and scrap the rest.

Excellent. I tried not to do that from the very beginning, because I was trying to preserve as much of the original code as possible (I didn't want to be the outsider who comes to break everything). You're going to say it wasn't necessary, but it's good to have your approval to make a deeper change haha.

I have some ideas I'd like to try that I think will help simplify the whole compiler. 

I'll keep you posted. Thanks!

**Comment by gzanitti** [comment]
Small update:

I'm almost done with the new code. The truth is: There are no big changes, just some reorganizations.

Roughly, these are the major changes:

- After the code is parsed, I take care of resolving each include and import, so that the code is as "expanded" as possible.
    - Biggest gain: This allows me to create a new assembler to handle the includes, and makes handling scopes easier. 
    - Biggest loss: After it comes out of the parser, I have to preprocess the code. I tried to do it all in one pass, but the code got too messy and I preferred to prioritize readability (at least for now, we can optimize it in the future). 
    - Important fact: This allowed me to discover that there are two tests where the expected values were wrong.

- I also make sure to save all macros before starting to assemble. Having all the macros predefined saves a lot of headaches, since calling a macro before declaring it means expanding part of the code in the future. Again, maybe it's not as efficient as it could be, but I prefer clarity over performance for now. Everything can be optimized in a future PR.
- Having all the macros already defined also helped me to simplify the assembler I had already defined to remove the pending state at the beginning of this PR, so there is no need to do inserts anymore, all new operations are placed at the end of the list, making the code much simpler and more understandable.

I still need to fix a few things. Most importantly, I would like to add some more tests that make use of the dynamic push of labels (%push(label)) to make sure everything is working properly. Also, the handling and resolution of paths in each include/import is just a "hacky patch". And there is a test that fails when the push label is an expression. Nothing that is likely to be a major headache, but I'd like to have it sorted out before you guys take a look at the code.


Anyway, I'm really enjoying working on this project, so I hope you like my ideas.

P.S.: I'm not a Rust expert, my code might not be "rusty" enough, so I'm open to any suggestions :).

P.S.2: I'm really starting to get attached to this project. Would you mind if I try to put together a logo and a small website? Of course I plan to open a thread where we can discuss ideas/designs. I'd really like to help spread the word about ETK. Hi Huff, I'm looking at you ;)

**Comment by gzanitti** [comment]
I think the code is ready :rocket: 
However, I'll be adding a few more tests over the next  days to be sure everything works perfectly.

Thanks!

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:214]: It's been a while, so bear with me, but how do we want to use "undeclared" vs "undefined"?

You're probably more familiar with the code than I am at this point, but IIRC, undefined labels were labels that had a `some_label:` in the source code, but didn't have a fixed position yet because of a `%push` macro. Undeclared labels were forward references to a `some_label:` on a later line.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:237]: If this is an internal function, it should be `pub(crate)`.

---

The Rust naming convention is to use something like `with_len` or `with_length` for a constructor that takes a length.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:250]: From a quick read, this looks like an implementation detail? Are you ever going to want to call `push_all` without inspecting the macros?

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:266]: Did you consider changing the API to just an `assemble` function? I don't think we need the `push_all`/`take` phases if we're only considering the whole program in one chunk.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:287]: Is it possible to un-nest the matches? Maybe something like:

```rust
Err(ops::Error::ContextIncomplete { source: UnknownLabel { .. } } => { /* ... */ },
Err(ops::Error::ContextIncomplete { source: UnknownMacro { name, .. } } => { /* ... */ },
Err(ops::Error::ContextIncomplete { source: UndefinedVariable { name, .. } } => { /* ... */ },
```

**Review by SamWilsn — COMMENTED** [review]
First few comments. Looking reasonable so far! I'm about halfway through `asm.rs`.

File: etk-asm/src/asm.rs
Comment (gzanitti) [reply etk-asm/src/asm.rs]: I was wrong about the two tests I thought were wrong. I fixed both tests and uploaded the assembler modification, so this function is no longer needed.

File: etk-asm/src/asm.rs
Comment (gzanitti) [reply etk-asm/src/asm.rs]: In this case, it's only necessary to keep track of the labels that were pushed before they were declared. I agree with you that the use of the word _undefined_ is not quite correct, so I made the modification so that the field is called _undeclared_labels_. I also corrected the documentation to better understand the meaning of each struct field.

File: etk-asm/src/asm.rs
Comment (gzanitti) [reply etk-asm/src/asm.rs]: I fully agree with this. I made the change from push_all to assemble and now the whole process is done through a single call. Even the creation of a new assembler to resolve scopes is now done internally, which makes much more sense.

Thanks!

**Comment by gzanitti** [comment]
> First few comments. Looking reasonable so far! I'm about halfway through `asm.rs`.

Thanks! Take your time and don't hesitate to ask questions or correct anything you think can be improved. 
Especially if something can be written in a more "rusty" way.

**Comment by gzanitti** [comment]
In addition to issue 124, mentioned at the beginning, this PR also fixes issue #108  (both tests mentioned in the issue were added to the test suite).

File: etk-asm/src/ingest.rs
Comment (SamWilsn) [inline etk-asm/src/ingest.rs:332]: This seems a bit crazy. Can't just do:

```suggestion
                                path: hex_path,
```

or maybe:

```suggestion
                                path: &hex_path,
```

?

File: etk-asm/src/ingest.rs
Comment (SamWilsn) [inline etk-asm/src/ingest.rs:339]: Yeah, there's something fishy about the way these conversions are happening.

File: etk-asm/src/ingest.rs
Comment (SamWilsn) [inline etk-asm/src/ingest.rs:353]: Generally in Rust, if you immediately clone a reference inside a function, you may as well take the owned type as a parameter. In this case, `PathBuf`.

That lets the caller decide whether to clone or not. If the caller no longer needs the variable, it can transfer ownership (saving the potentially expensive clone.)

File: etk-asm/src/ingest.rs
Comment (SamWilsn) [inline etk-asm/src/ingest.rs:360]: Is the turbofish (`::<>`) necessary here?

File: etk-asm/src/parse/error.rs
Comment (SamWilsn) [inline etk-asm/src/parse/error.rs:67]:
```suggestion
    #[snafu(display("File `{}` does not exist", path.to_string_lossy()))]
```

File: etk-asm/src/parse/error.rs
Comment (SamWilsn) [inline etk-asm/src/parse/error.rs:69]: What returns this variant?

File: etk-asm/src/parse/error.rs
Comment (SamWilsn) [inline etk-asm/src/parse/error.rs:80]: Similarly, what returns this variant?

**Review by SamWilsn — COMMENTED** [review]
Another partial review. Sorry!

File: etk-asm/src/ingest.rs
Comment (gzanitti) [reply etk-asm/src/ingest.rs]: To be honest, the compiler asked me to add it so that the code can compile.  But I just removed it as you said and now there is no error. Possibly related to my first implementation. I'm sorry

File: etk-asm/src/parse/error.rs
Comment (gzanitti) [reply etk-asm/src/parse/error.rs:69]: I don't understand this comment, what do you mean?

File: etk-asm/src/ingest.rs
Comment (gzanitti) [reply etk-asm/src/ingest.rs]: This is rare. This was, after quite a while, the only way I could get the code to work. But I just removed it and it just works :open_mouth:

File: etk-asm/src/ingest.rs
Comment (gzanitti) [reply etk-asm/src/ingest.rs]: Depth is left over from my first implementation of "Program" when I was only keeping track of the latest source. I have just removed it

File: etk-asm/src/ingest.rs
Comment (gzanitti) [reply etk-asm/src/ingest.rs]: I simplified the code a bit so that push_path gets a reference and made some changes to bring the clone to the one place where it was indispensable. Sorry, while I've made a lot of progress in this time, there are still some "common practices" of Rust that escape me.

**Comment by gzanitti** [comment]
> Another partial review. Sorry!

Thank you for your comments @SamWilsn. I just hope I'm not giving you too many headaches haha.

P.S: You can take your time on the review, in the meantime I'm finishing PR #135  and, if necessary, I can spend some time on solving some other items on the list of issues in the meantime.

**Comment by SamWilsn** [comment]
> P.S.2: I'm really starting to get attached to this project. Would you mind if I try to put together a logo and a small website? Of course I plan to open a thread where we can discuss ideas/designs. I'd really like to help spread the word about ETK. Hi Huff, I'm looking at you ;)

I missed this part! I'd absolutely love that! There are so many little things I've wanted to build with ETK, and I had basically given up on having the time/motivation to do it. Extremely happy to have someone as excited as you. A website (maybe with an ETK-compiled-to-JavaScript playground :innocent:) is a great idea!

File: etk-asm/src/parse/error.rs
Comment (SamWilsn) [reply etk-asm/src/parse/error.rs:69]: If I search through your branch, I only find this definition. I don't see the variant `FileNotFound` being used anywhere. It's entirely possible I'm just missing where it's used, but I wanted to double check before merging.

File: etk-asm/src/ingest.rs
Comment (SamWilsn) [reply etk-asm/src/ingest.rs]: Don't be sorry, that's what reviews are for!

File: etk-asm/src/ingest.rs
Comment (SamWilsn) [reply etk-asm/src/ingest.rs]: I only mentioned it because of your postscript. It's such a tiny thing that I wouldn't've mentioned it otherwise.

File: etk-asm/src/ingest.rs
Comment (SamWilsn) [reply etk-asm/src/ingest.rs]: snafu sometimes amazes me, and sometimes confounds me.

File: etk-asm/src/ingest.rs
Comment (gzanitti) [reply etk-asm/src/ingest.rs]: Don't worry, these comments are really helpful. There are many things you don't learn in books/tutorials.

**Comment by gzanitti** [comment]
> > P.S.2: I'm really starting to get attached to this project. Would you mind if I try to put together a logo and a small website? Of course I plan to open a thread where we can discuss ideas/designs. I'd really like to help spread the word about ETK. Hi Huff, I'm looking at you ;)
> 
> I missed this part! I'd absolutely love that! There are so many little things I've wanted to build with ETK, and I had basically given up on having the time/motivation to do it. Extremely happy to have someone as excited as you. A website (maybe with an ETK-compiled-to-JavaScript playground 😇) is a great idea!

Great, glad to be able to help, and to discuss and develop all the ideas you have in mind. I'll prepare a prototype in the next few weeks, including the playground you mention, and I'll write to you to see what you think :smile:

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:258]: This looks like it discards errors from `concretize_ops`.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:326]: I think at this point we can just use a slice instead of an iterator.

```suggestion
    pub fn assemble<O>(&mut self, ops: &[O]) -> Result<Vec<u8>, Error>
```

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:328]:
```suggestion
```

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:260]: Let's just move this directly into `assemble`.

```suggestion
```

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:339]: Moving from `take`:

```suggestion
        let output = self.concretize_ops()?;
        self.ready.clear();
        Ok(output)
```

File: etk-asm/src/asm.rs
Comment (SamWilsn) [reply etk-asm/src/asm.rs:240]: Ah, yeah, I see now. You're doing two passes, so you need to clone the iterator. I suggested changing `assemble` to just take a slice, so no more cloneable iterator.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:379]: Is it possible to combine `push` with `push_rawop`? Like move `self.expand_macro(&m.name, &m.parameters)?;` (line 373) to line 424?

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:424]: If I understand this correctly, hitting a macro invocation inside this function is a bug, right? We should either merge `push` and `push_rawop`, or put `unreachable!()` here instead of `Ok(())`.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:450]:
```suggestion
                        source: UnknownLabel { .. },
```

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:456]: If `op` is unsized, does it make sense to add two here? If I'm assembling something like:

```
%push(faraway)
%repeat_jumpdest(511)
faraway:
```

The push would have to be three bytes long.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:453]: Why do you use `size` to determine if this is a variable push? A more accurate way would be something like:

```rust
if let AbstractOp::Push(_) = op { ... }
```

File: etk-asm/src/asm.rs
Comment (gzanitti) [reply etk-asm/src/asm.rs:456]: Yes, exactly, but in this instance it works more like a placeholder. Then, when the operations are actually concretised and the final position of the label is known, it is corrected. That's why also the dynamic jump flag is set to true. 

In fact, there is a test to prove that this works: 
https://github.com/gzanitti/etk/blob/12c609c081bc96812b067b12a45192c1f700e3c9/etk-asm/tests/asm.rs#L297

File: etk-asm/src/asm.rs
Comment (gzanitti) [reply etk-asm/src/asm.rs:453]: My idea was to make it as general as possible. I was not only interested in pushes, but in any other operation that won't have a defined size in the future. But rethinking it, you'll probably have to modify other parts of the code if that happens, so the idea of using if let AbstractOp::Push(_) makes a lot of sense. Thanks!

File: etk-asm/src/asm.rs
Comment (gzanitti) [reply etk-asm/src/asm.rs:424]: Not really. While they have all been processed at this point, they are not removed from the original code. So this is simply to ignore them.

Comment (gzanitti) [reply etk-asm/src/asm.rs:424]: Yes, it's not exactly a bug, but since the macro had already been processed before, it was simply ignored. I will combine both pushes to simplify the code.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:371]: Why use floating point for this? It seems equivalent to `(self.concrete_len - ul.position) / 256`. Integer division in Rust always floors.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:377]: Something smells off about this, but I can't quite put my finger on it. Let me try a few things, and I'll be back.

Comment (gzanitti) [reply etk-asm/src/asm.rs:377]: Sure! Let me know if you have any questions or if I can help you in any way. Basically that part of the code takes care of compensating the value of the labels. If a dynamic push was used to define the label in question, in that instance its position is already known, so it "stretches" the value to reach the final position. As all intermediate macros were already expanded because they are known beforehand, there should be no problem in setting their value correctly.

But if you find any interesting edge case, feel free to let me know and I can add the test to be sure it works.

Comment (gzanitti) [reply etk-asm/src/asm.rs:377]: Oh, I see, I didn't realise that labels could be defined that way. Excellent catch. I'll fix the code as quickly as possible.

Comment (SamWilsn) [reply etk-asm/src/asm.rs]: My thought is that these two programs must assemble to the same thing:

<table>
<tr>
<th>
</th>
<th>Constant Push</th>
<th>Variable Push</th>
</tr>
<tr>
<th>Input</th>
<td>

```
push2 label + 254                   
label:
pc
```

</td>
<td>

```
%push(label + 254)                   
label:
pc
```

</td>
</tr>

<tr>
<th>Output</th>
<td>

```
61010158
```

</td>
<td>

```
61010058
```

</td>
</tr>
</table>

I assumed the constant push was correct, but regardless, they should be the same.

Comment (gzanitti) [reply etk-asm/src/asm.rs]: Oh, I see. I'll check it out. Thanks!

Comment (gzanitti) [reply etk-asm/src/asm.rs]: Just to be sure. This should be a valid program too, right?

```
%push(label1 + label2 + 254)
label:
pc
label2:
pc
```

Comment (gzanitti) [reply etk-asm/src/asm.rs]: Nothing out of the ordinary; I was just working on the previous case and noticed that this one wasn't included in my solution, and I thought it should be. I wanted to double-check. I'll figure out a solution that covers both, but it might mean changing some parts you've already looked at. Sorry about that.

Comment (gzanitti) [reply etk-asm/src/asm.rs]: Solved. There were many intermediate steps, so I simplified the code a bit and renamed things for better understanding.

File: etk-asm/src/asm.rs
Comment (gzanitti) [reply etk-asm/src/asm.rs]: Hi Sam. I improved a little the documentation of the function and added some comments so that you can better understand what's happening in each step. I hope it's clearer now :)

**Comment by SamWilsn** [comment]
Sorry it's taken me so long to re-review! I believe I found another case that passed before this pull request and is failing now.

Thanks again for being so persistent on this. I _really_ appreciate the help getting ETK into a good place <3

**Comment by gzanitti** [comment]
> Sorry it's taken me so long to re-review! I believe I found another case that passed before this pull request and is failing now.
> 
> Thanks again for being so persistent on this. I _really_ appreciate the help getting ETK into a good place <3

Sorry, it's also my fault for having to change several parts of the code in the middle of the process and make you check everything again.  As I always say, thanks for your patience and for letting me collaborate :)

Regarding the code, I solved the case you mention and also added some more tests. The novelty is that now all the pushes are checked and updated taking into account the previous updates (acum variable) (declared_labels is now a BTreeMap because it matters the order when adding the accumulated size to the labels).

**Comment by gzanitti** [comment]
Hey @SamWilsn, I needed a HashMap sorted by insert and ended up adding IndexMap to the project. However, it is only used in the assembler and could be replaced by a struct with a HashMap and a Vector if you don't want to add more dependencies. 

Let me know what you think.

**Review by SamWilsn — COMMENTED** [review]
Uggh, I found another test case that fails with this approach. I'll push it up in a second. I'm so sorry x.x

**Comment by SamWilsn** [comment]
I'm totally happy with the `IndexMap`.

**Comment by SamWilsn** [comment]
Pushed the new test. Summary is:

<table>
<tr><td></td><th>Variable + Fixed</th><th>Variable + Variable</th></tr>
<tr>
<th>Source</th>

<td>

```
%push(lbl1 - lbl2)
push2 lbl1 + lbl2
pc # repeat 126 times.
lbl1:
lbl2:
```

</td>
<td>

```
%push(lbl1 - lbl2)
%push(lbl1 + lbl2)
pc # repeat 126 times.
lbl1:
lbl2:
```

</td>
</tr>
<tr>
<th>Outcome</th>
<td>

`6000610106...`

</td>
<td>

`6000610108...`

</td>
</tr>
</table>

**Comment by gzanitti** [comment]
Hey @SamWilsn , how are you?
First of all, sorry for the delay. I've been having some personal issues and couldn't dedicate as much time as I would have liked to ETK. Also, my country is slowly going into the abyss, which doesn't help.

Don't worry about mentioning as many corrections as you deem necessary. After all, this code belongs more to you than to me and I can make all the changes you consider necessary until you are satisfied :).

Regarding the bug you reported, I'm sorry I didn't see it before, I understand that with the last change nothing like this should happen again.

I look forward to your comments!

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:400]: Can we use [`BigInt::bits`](https://docs.rs/num-bigint/latest/num_bigint/struct.BigInt.html#method.bits) and [integer ceiling division](https://stackoverflow.com/a/2745086) here instead of converting to a float? I'm not sure if this'll give the correct answer for really big numbers.

File: etk-asm/src/asm.rs
Comment (SamWilsn) [inline etk-asm/src/asm.rs:403]: Would [`IndexMap::values_mut`](https://docs.rs/indexmap/latest/indexmap/map/struct.IndexMap.html#method.values_mut) be more appropriate?

**Review by SamWilsn — COMMENTED** [review]
Just these last two items!

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**
